mittab/apps/tab/tab_card.py

```diff
@@ -0,0 +1,284 @@
+        ]
+    dstat2 = []
+    if not iron_man:
+        dstat2 = [
+            k for k in RoundStats.objects.filter(debater=deb2).filter(
+                round=round_obj).all()
+        ]
+    blank_rs = RoundStats(debater=deb1, round=round_obj, speaks=0, ranks=0)
+    while len(dstat1) + len(dstat2) < 2:
+        # Something is wrong with our data, but we don't want to crash
+        dstat1.append(blank_rs)
+    if not dstat2 and not dstat1:
+        return None, None
+    if not dstat2:
+        dstat1, dstat2 = dstat1[0], dstat1[1]
+    elif not dstat1:
+        dstat1, dstat2 = dstat2[0], dstat2[1]
+    else:
+        dstat1, dstat2 = dstat1[0], dstat2[0]
+    return dstat1, dstat2
+
+def json_get_round(round_obj, team, deb1, deb2):
+    jsonRound = {
+        "round_number": round_obj.round_number,
+        "round_id": round_obj.pk,
+        "side": "G" if round_obj.gov_team == team else "O",
+        "result": get_victor_label(round_obj.victor, "G" if round_obj.gov_team == team else "O"),
+        "chair": round_obj.chair.name,
+        "wings": [judge.name for judge in round_obj.judges.exclude(pk=round_obj.chair.pk)]
+    }
+    
+    opponent = round_obj.opp_team if round_obj.gov_team == team else round_obj.gov_team
+    if opponent:
+        opponent_debaters = list(opponent.debaters.all())
+        jsonRound["opponent"] = {
+            "name": opponent.get_team_code(),
+            "school": opponent.school.name,
+            "debater1": opponent_debaters[0].name if opponent_debaters else None,
+            "debater2": opponent_debaters[1].name if len(opponent_debaters) > 1 else None,
+        }
+
+    jsonRound["debater1"] = list(
+        (stat.speaks, stat.ranks) 
+        for stat in RoundStats.objects.filter(debater=deb1, round=round_obj)
+    )
+    jsonRound["debater2"] = list(
+        (stat.speaks, stat.ranks) 
+        for stat in RoundStats.objects.filter(debater=deb2, round=round_obj)
+    ) if deb2 else []
+
+    try:
+        bye_round = Bye.objects.get(bye_team=team).round_number
+        jsonRound[bye_round - 1][bye_round]
+    except Bye.DoesNotExist:
+        pass
+
+    return jsonRound
+
+class JSONDecimalEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, Decimal):
+            return float(obj)
+        return json.JSONEncoder.default(self, obj)
+    
+def get_all_json_data():
+    all_tab_cards_data = {}
+    
+    # Prefetch related data to reduce database queries
+    teams = Team.objects.prefetch_related(
+        'school',
+        Prefetch('debaters', queryset=Debater.objects.all()),
+        Prefetch('gov_team', queryset=Round.objects.all()),
+        Prefetch('opp_team', queryset=Round.objects.all())
+    )
+
+    total_rounds = TabSettings.objects.get(key="tot_rounds").value
+
+    for team in teams:
+        tab_card_data = {
+            "team_name": team.get_team_code(),
+            "team_school": team.school.name,
+            "rounds": [{}] * total_rounds
```

Comment:
This is a pretty minor stylistic nit, but I don't love generating `rounds` just to mutate it -- we're basically using this initial value just as a way to indicate how many rounds to fill

```python3
tab_card_data = { "team_name", team.get_team_code(), "team_school: team.school }

# ...

# Treat this as a mapping for round lookup rather
# than the source-of-truth on number of rounds
rounds_by_zero_based_round_number = [None] * total_rounds
for round_obj in rounds:
  # side note: I actually totally forget if "Round 1" means round_number=0 or round_number=1. need to verify
  rounds_by_round_number[round_obj.round_number - 1] = round_obj

all_tab_card_data["rounds"] = map(
  lambda zero_based_round_num:
    round_for_round_num = rounds_by_zero_based_round_number[zero_based_round_num]
    return json_get_round(round_obj, team, deb1, deb2),
  rounds
)
```

Reply:
Not sure if this fully answers the stylistic concern, but I think since the `json_get_round` data itself contains the round numbers, the cleanest option here might just be getting rid of the correspondence between the index of the array and the round number, since the same data is available within the json. Similarly, we could also replace the array with a dictionary
```
rounds : {
    1 : [round 1 data],
    3 : [round 3 data]
```

Either way I think is a bit cleaner than populating `None` or `{}` for every missing round. For the commit I'm pushing soon I did:

```
for team in teams:
        tab_card_data = { "team_name": team.get_team_code(), "team_school": team.school }

        #...
        rounds = list(team.gov_team.all()) + list(team.opp_team.all())

        round_data = []
        for round_obj in rounds:
            round_data.append(json_get_round(round_obj, team, deb1, deb2))
        tab_card_data["rounds"] = round_data

```